sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,500 @@
+      MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
+
+      ArrayList splitKeys = null;
+      if (numSplits > 0) {
+        // the user defines his desired number of splits
+        // calculate the bundle size
+        long estimatedSizeBytes = getEstimatedSizeBytes(options);
+        desiredBundleSizeBytes = estimatedSizeBytes / numSplits;
+      }
+
+      // now we have the desized bundle size (provided by user or provided by the runner)
+      // we use Mongo splitVector command to get the split keys
+      BasicDBObject splitVectorCommand = new BasicDBObject();
+      splitVectorCommand.append("splitVector", database + "." + collection);
+      splitVectorCommand.append("keyPattern", new BasicDBObject().append("_id", 1));
+      splitVectorCommand.append("force", false);
+      // maxChunkSize is the Mongo partition size in MB
+      splitVectorCommand.append("maxChunkSize", desiredBundleSizeBytes / 1024 / 1024);
+      Document splitVectorCommandResult = mongoDatabase.runCommand(splitVectorCommand);
+      splitKeys = (ArrayList) splitVectorCommandResult.get("splitKeys");
+
+      return createSourceList(splitKeys);
+    }
+
+    /**
+     * Create a sources list. Each source in the list has a subset of keys to deal with (split).
+     *
+     * @param splitKeys List of split keys to create sources.
+     * @return The list of sources.
+     */
+    private List<BoundedSource> createSourceList(ArrayList splitKeys) {
+      List<BoundedSource> splitSourceList = new ArrayList<>();
+      if (splitKeys == null) {
+        splitSourceList.add(new BoundedMongoDbSource(uri, database, collection, filter, numSplits));
+      } else {
+        String lastID = null; // lower boundary of the first min split
+        int listIndex = 0;
+        for (final Object splitKey : splitKeys) {
+          String currentID = splitKey.toString();
+          String filterWithBoundaries;
+          if (filter != null && !filter.isEmpty()) {
+            if (listIndex == 0) {
+              filterWithBoundaries = "{ $and: [ {\"_id\":{$lte:ObjectId(\""
+                  + currentID + "\")}}, "
+                  + filter + " ]}";
+            } else if (listIndex == (splitKeys.size() - 1)) {
+              filterWithBoundaries = "{ $and: [ {\"_id\":{$gt:ObjectId(\"" + lastID
+                  + "\")," + "$lt:ObjectId(\"" + currentID
+                  + "\")}}, " + filter + " ]}";
+              splitSourceList.add(new BoundedMongoDbSource
+                  (uri, database, collection, filterWithBoundaries, numSplits));
+              filterWithBoundaries = "{ $and: [ {\"_id\":{$gt:ObjectId(\"" + currentID
+                  + "\")}}, " + filter + " ]}";
+            } else {
+              filterWithBoundaries = "{ $and: [ {\"_id\":{$gt:ObjectId(\"" + lastID
+                  + "\")," + "$lte:ObjectId(\"" + currentID + "\")}}, "
+                  + filter + " ]}";
+            }
+          } else {
```

Comment:
Seems like you can build a string for the ObjectId filter, and then optionally also wrap it into an $and with the specified filter (if any). That way you'll have fewer code duplication between lines 317-332 and 334-345.